kms.go

```diff
@@ -0,0 +1,186 @@
+package hdfs
+
+import (
+	"bytes"
+	"crypto/aes"
+	"io/ioutil"
+
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	hdfs "github.com/colinmarc/hdfs/v2/internal/protocol/hadoop_hdfs"
+	spnego "github.com/jcmturner/gokrb5/v8/spnego"
+)
+
+const (
+	kmsSchemeHTTP  = "kms://http@"
+	kmsSchemeHTTPS = "kms://https@"
+)
+
+func (c *Client) kmsAuth(url string) error {
+	if c.options.KerberosClient == nil {
+		url += ("&user.name=" + c.options.User)
+	}
+
+	req, err := http.NewRequest("OPTIONS", url, nil)
+	if err != nil {
+		return err
+	}
+
+	var resp *http.Response
+	if c.options.KerberosClient != nil {
+		kHttp := spnego.NewClient(c.options.KerberosClient, c.http, "")
+		resp, err = kHttp.Do(req)
+	} else {
+		resp, err = c.http.Do(req)
+	}
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("bad kms auth status: %v", resp.StatusCode)
+	}
+	return nil
+}
+
+func kmsSplit(keyProvider string) []string {
+	return strings.Split(keyProvider, ",")
+}
+
+// kmsUrl parse KeyProviderUri to list of URL's
+func (c *Client) kmsUrl(einfo *hdfs.FileEncryptionInfoProto) ([]string, error) {
+	defaults, err := c.fetchDefaults()
+	if err != nil {
+		return nil, err
+	}
+
+	uri := defaults.GetKeyProviderUri()
+	if uri == "" {
+		return nil, errors.New("KeyProviderUri not configured on server")
+	}
+
+	var urls []string
+	var proto string
+	if strings.HasPrefix(uri, kmsSchemeHTTPS) {
+		proto = "https://"
+		urls = kmsSplit(uri[len(kmsSchemeHTTPS):])
```

Comment:
> The hostnames are separated by semi-colons, and all KMS instances must run on the same port.
via: https://hadoop.apache.org/docs/stable/hadoop-kms/index.html

but `kmsSplit` actually split by `,`. Port number must be carefully handled too.

Reply:
fixed.